### DISCUSSION TITLE
关于Service Worker的设置问题

### DISCUSSION THREAD

**Discussion by MCSeekeri** [volantis-x/hexo-theme-volantis#791] [Q&A] [answered]
文档内的Service Workers设置部分让我有点摸不着头脑……因为两个PWA插件都有点老，gulp似乎是唯一选择……
但是每次更改都要修改文件就有点麻烦了，能基于gulp自动生成的话最好。

我其实想要实现类似于目前主题站一样的volantis-sw.js，但我个人能力不足，不会写……

所以哪位可以给我指个路？我已经搜索过一些文章，但是实际上手基本上都会有或多或少的错误……

**Comment by Tikas** [comment]
gulp 并不是缓存插件，它只是压缩工具

文档里的两个 Service Workers 插件，本身项目里有使用方法介绍。

Volantis 本身的 sw.js，你可以来这里查看源代码：[点我前往 Volantis 官网项目查看 sw.js](https://github.com/volantis-x/community/blob/main/source/volantis-sw.js)

由于我不懂 js ，没去进一步分析，具体如何实现，这还需要他人或您自己学习研究了

**Comment by Tikas** [comment]
Gulp 完全可以写到相关命令执行后再执行它，这样可以达到自动生成的效果

具体实现过程举例：

先清除缓存：hexo clean 

之后再构建，此时命令由上面变成：hexo clean && hexo generate，这条命令是既执行了前面的清除缓存，又执行了这次的构建

接下来，可以再执行下一步操作：hexo clean && hexo generate && gulp，这条命令是执行了清除缓存，再执行构建，然后再执行 gulp 压缩

如果还需要发布，可以再在后面加 hexo deploy，当然，是否需要看个人使用场景

如果懒，不想输入这么多命令，完全可以把这些命令集合到项目的脚本里：

在项目文件：package.json 里面的  scripts 新加自己喜欢的命令执行指令，增加上面的执行步骤，就可以了，举例：

```json
"scripts": {
    "start": "hexo clean && hexo g && gulp",
    "g": "hexo clean && hexo generate",
    "c": "hexo clean",
    "d": "hexo deploy",
    "s": "hexo clean && hexo server"
  },
```
以上是直接复制 volantis 官网的，完整参考，[请点此链接查看](https://github.com/volantis-x/community/blob/main/package.json)

**Reply by Tikas** [reply]
使用方法：npm run start

如果你是使用 yarn，那么使用方法就是 yarn start

**Reply by MCSeekeri** [reply]
抱歉让您花了这么多时间，但我已经给我的网站配置好gulp了。
CI也做好了相应的设置。
参考[Repo](https://github.com/mcseekeri/sciadv)

**Reply by Tikas** [reply]
不用客气！

**Comment by MHuiG** [comment, marked as answer]
官网的相关文件：

https://github.com/volantis-x/community/blob/main/source/volantis-sw.js

https://github.com/volantis-x/community/blob/main/.github/npm-version-bump.js

https://github.com/volantis-x/community/blob/main/.github/cache.py

https://github.com/volantis-x/community/blob/main/.github/workflows/auto-deploy.yml

https://github.com/volantis-x/community/blob/main/source/_volantis/headBegin.ejs

**Reply by MHuiG** [reply]
https://www.npmjs.com/package/@mhg/volantis-community

**Reply by MCSeekeri** [reply]
这里还有一个小问题，就是如果站点部署了百度和谷歌的统计代码时似乎代码不会加载……
隔壁ClientWorker使用了skip来跳过，但这个SW应该怎么处理……
我部署的位置[在此](https://sciadv-git-test-mcseekeri.vercel.app/)

**Reply by MHuiG** [reply]
写个正则表达式匹配统计器发出的请求调用NetworkOnly

https://github.com/volantis-x/community/blob/main/source/volantis-sw.js#L18

**Reply by MCSeekeri** [reply]
> 写个正则表达式匹配统计器发出的请求调用NetworkOnly
> 
> https://github.com/volantis-x/community/blob/main/source/volantis-sw.js#L18

Σ(っ °Д °;)っ
秒回！
我去写写试试

**Reply by MCSeekeri** [reply]
![image](https://user-images.githubusercontent.com/20928094/181397345-ddadb56b-0c36-412d-bb87-d34e7a15ed37.png)
我还是搞不大懂……因为我不会写JS……

**Reply by lete114** [reply]
芜湖

**Reply by ChenYFan** [reply]
NetworkOnly其实就是透明请求，在外部看就是不经过sw处理，和cw的skip其实一个道理，解决莫名其妙的兼容性问题时很方便

**Comment by MHuiG** [comment]
另外推荐 
https://github.com/ChenYFan/ClientWorker

https://github.com/Lete114/SW-Racing

这两个如果有问题直接私聊 @Lete114 和 @ChenYFan

**Reply by MCSeekeri** [reply]
好，这就去群里烦他俩（

**Comment by inkss** [comment]
试试这个  https://www.pwabuilder.com/

**Reply by MCSeekeri** [reply]
好的，我之后看一眼。
不过我已经配置完一个很简陋的PWA了（

**Comment by MCSeekeri** [comment]
新的问题出现了，因为更新SW需要一点点时间，还需要刷新页面，这些操作对用户来讲是不可见的……
所以有没有那种提示站点更新，刷新以载入新内容的东西（
我很菜所以不会写（
近似的参考是Butterfly和Waline的文档。

**Reply by Tikas** [reply]
之前官网里有几个文章是写弹窗提示的，刚刚好有 sw 的部分

比如 https://goopher.tk/

**Comment by xingpingcn** [comment]
推荐一下我自己用的sw：https://github.com/xingpingcn/picture-bed-using-freecdn
